samples/net/cloud/aws_iot_mqtt/src/main.c

```diff
@@ -0,0 +1,494 @@
+/*
+ * Copyright (c) 2023 Lucas Dietrich <ld.adecy@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "creds/creds.h"
+#include "dhcp.h"
+
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+#include <poll.h>
+#include <unistd.h>
+
+#include <zephyr/data/json.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/net/dns_resolve.h>
+#include <zephyr/net/mqtt.h>
+#include <zephyr/net/sntp.h>
+#include <zephyr/net/tls_credentials.h>
+#include <zephyr/random/rand32.h>
+
+
+#if defined(CONFIG_MBEDTLS_MEMORY_DEBUG)
+#include <mbedtls/memory_buffer_alloc.h>
+#endif
+
+#include <sys/socket.h>
+LOG_MODULE_REGISTER(aws, LOG_LEVEL_DBG);
+
+#define SNTP_SERVER "0.pool.ntp.org"
+
+#define AWS_BROKER_PORT "8883"
+
+#define MQTT_BUFFER_SIZE 256u
+#define APP_BUFFER_SIZE	 4096u
+
+#define MAX_RETRIES	    10u
+#define BACKOFF_EXP_BASE_MS 1000u
+#define BACKOFF_EXP_MAX_MS  60000u
+#define BACKOFF_CONST_MS    5000u
+
+static struct sockaddr_in aws_broker;
+
+static uint8_t rx_buffer[MQTT_BUFFER_SIZE];
+static uint8_t tx_buffer[MQTT_BUFFER_SIZE];
+static uint8_t buffer[APP_BUFFER_SIZE]; /* Shared between published and received messages */
+
+static struct mqtt_client client_ctx;
+
+static const char mqtt_client_name[] = CONFIG_AWS_THING_NAME;
+
+static uint32_t messages_received_counter;
+static bool do_publish;	  /* Trigger client to publish */
+static bool do_subscribe; /* Trigger client to subscribe */
+
+#define TLS_TAG_DEVICE_CERTIFICATE 1
+#define TLS_TAG_DEVICE_PRIVATE_KEY 1
+#define TLS_TAG_AWS_CA_CERTIFICATE 1
+
+static sec_tag_t sec_tls_tags[] = {
+	TLS_TAG_DEVICE_CERTIFICATE,
```

Comment:
Ah, I've started to wonder how does it work if you only provide device certificate tag and no root CA, but now I see they're all regsitered under the same tag number.

I think it'd be a bit more readable, if root CA received its own tag, as it's not related to the device credentials. Device private tag and cert should share the same tag though, as they come in couple.